api/dataprocessor_test.go

```diff
@@ -1095,6 +1096,227 @@ func TestMergeSeries(t *testing.T) {
+	chunkSpan := uint32(600)
+	numChunks := uint32(10)
+	stats, _ := helper.New(false, "", "standard", "metrictank", "")
+	srv, _ := NewServer(stats)
+	store := mdata.NewMockStore()
+	srv.BindBackendStore(store)
+	metrics := mdata.NewAggMetrics(store, chunkSpan, numChunks, 0, 0, 0, 0, []mdata.AggSetting{})
+	mdata.InitMetrics(stats)
+	srv.BindMemoryStore(metrics)
+	defer metrics.Stop()
+	c := cache.NewCCache()
+	srv.BindCache(c)
+	metricKey := "metric1"
+
+	// generating 4 chunks, 3 will go into the cache and 1 into the backend store.
+	// this simulates a scenario where the latest chunk has been persisted but not
+	// queried (and cached) yet.
+	chunks := make([]*chunk.Chunk, 0)
+	for i := chunkSpan; i < chunkSpan*5; i++ {
+		if i%chunkSpan == 0 {
+			chunks = append(chunks, chunk.New(i))
+		}
+		chunks[len(chunks)-1].Push(i, float64(i*2))
+	}
+
+	prevts := uint32(0)
+	// adding 3 itgens to cache
+	var i int
+	for i = 0; i < 3; i++ {
+		chunks[i].Series.Finish()
+		itgen := chunk.NewBareIterGen(chunks[i].Series.Bytes(), chunks[i].Series.T0, chunkSpan)
+		srv.Cache.Add(metricKey, prevts, *itgen)
+		prevts = itgen.Ts()
+	}
+
+	// preparing the next itgen as store mock result (to simulate one that has just been added and not cached yet)
+	chunks[i].Series.Finish()
+	itgens := []chunk.IterGen{*chunk.NewBareIterGen(chunks[i].Series.Bytes(), chunks[i].Series.T0, chunkSpan)}
+
+	// save some electrons by skipping steps that are no edge cases
+	steps := chunkSpan / 10
+
+	// we want to have:
+	// - queries that only query chunks in the cache
+	// - queries that only query chunks in cassandra
+	// - queries that query across multiple chunks in the cache
+	// - queries that query across chunks in the cache and cassandra
+	// so we can be sure that getSeriesCachedStore correctly stitches the returned results from different sources together
+	for from := uint32(chunkSpan + chunkSpan/2); from < uint32(chunkSpan*4); from += steps {
+		for to := from + 1; to <= from+chunkSpan+1; to += steps {
+			// resetting mock at the beginning of each iteration
+			store.ResetMock()
+			store.AddMockResult(itgens, nil)
+
+			req := reqRaw(metricKey, from, to, chunkSpan, 1, consolidation.None)
+			req.ArchInterval = 1
+			ctx := newRequestContext(&req, consolidation.None)
+			iters := srv.getSeriesCachedStore(ctx, to)
+
+			// expecting the first returned timestamp to be the T0 of the chunk containing "from"
+			expectResFrom := from - (from % chunkSpan)
+
+			// expecting the last returned timestamp to be the last point in the chunk containing "to"
+			expectResTo := (to - 1) + (chunkSpan - (to-1)%chunkSpan) - 1
+
+			// for each timestamp in the returned iterators we compare if it has the expected value
+			// we use the valueTracker to increase together with the iterators and compare at each step
+			valueTracker := expectResFrom
+
+			ts := make([]uint32, 0)
+			for _, it := range iters {
+				for it.Next() {
+					val, _ := it.Values()
+					if val != valueTracker {
```

Comment:
While it's not techncially incorrect to refer to the timestamps as values here (since they are "values" that we validate),
the terminology is confusing, as what we retrieve from it.Values(), and what we verify, are timestamps, and we ignore the "values" from the iterator. so i would s/val/ts/ and s/valueTracker/expectedTs/ which would be a more consistent name compared to the other expected* variables. (PS sorry to be nitpicky!)